File: duplicates-android/app/src/main/java/com/github/duplicates/message/MessageProvider.java
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.message;

import android.Manifest;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.Telephony;

import com.github.duplicates.DuplicateProvider;

import static android.provider.BaseColumns._ID;
import static android.provider.Telephony.TextBasedSmsColumns.ADDRESS;
import static android.provider.Telephony.TextBasedSmsColumns.BODY;
import static android.provider.Telephony.TextBasedSmsColumns.DATE;
import static android.provider.Telephony.TextBasedSmsColumns.DATE_SENT;
import static android.provider.Telephony.TextBasedSmsColumns.ERROR_CODE;
import static android.provider.Telephony.TextBasedSmsColumns.LOCKED;
import static android.provider.Telephony.TextBasedSmsColumns.PERSON;
import static android.provider.Telephony.TextBasedSmsColumns.PROTOCOL;
import static android.provider.Telephony.TextBasedSmsColumns.READ;
import static android.provider.Telephony.TextBasedSmsColumns.SEEN;
import static android.provider.Telephony.TextBasedSmsColumns.STATUS;
import static android.provider.Telephony.TextBasedSmsColumns.SUBJECT;
import static android.provider.Telephony.TextBasedSmsColumns.THREAD_ID;
import static android.provider.Telephony.TextBasedSmsColumns.TYPE;

/**
 * Provide duplicate messages.
 *
 * @author moshe.w
 */
public class MessageProvider extends DuplicateProvider<MessageItem> {

    private static String[] PERMISSIONS_READ = {Manifest.permission.READ_SMS};
    private static String[] PERMISSIONS_WRITE = {"android.permission.WRITE_SMS"};

    private static final String[] PROJECTION = {
            _ID,
            ADDRESS,
            BODY,
            DATE,
            DATE_SENT,
            ERROR_CODE,
            LOCKED,
            PERSON,
            PROTOCOL,
            READ,
            SEEN,
            STATUS,
            SUBJECT,
            THREAD_ID,
            TYPE
    };

    private static final int INDEX_ID = 0;
    private static final int INDEX_ADDRESS = 1;
    private static final int INDEX_BODY = 2;
    private static final int INDEX_DATE_RECEIVED = 3;
    private static final int INDEX_DATE_SENT = 4;
    private static final int INDEX_ERROR_CODE = 5;
    private static final int INDEX_LOCKED = 6;
    private static final int INDEX_PERSON = 7;
    private static final int INDEX_PROTOCOL = 8;
    private static final int INDEX_READ = 9;
    private static final int INDEX_SEEN = 10;
    private static final int INDEX_STATUS = 11;
    private static final int INDEX_SUBJECT = 12;
    private static final int INDEX_THREAD_ID = 13;
    private static final int INDEX_TYPE = 14;

    public MessageProvider(Context context) {
        super(context);
    }

    @Override
    protected Uri getContentUri() {
        return Telephony.Sms.CONTENT_URI;
    }

    @Override
    protected String[] getCursorProjection() {
        return PROJECTION;
    }

    @Override
    public MessageItem createItem(Cursor cursor) {
        return new MessageItem();
    }

    @Override
    public void populateItem(Cursor cursor, MessageItem item) {
        item.setId(cursor.getLong(INDEX_ID));
        item.setAddress(empty(cursor, INDEX_ADDRESS));
        item.setBody(empty(cursor, INDEX_BODY));
        item.setDateReceived(cursor.getLong(INDEX_DATE_RECEIVED));
        item.setDateSent(cursor.getLong(INDEX_DATE_SENT));
        item.setErrorCode(cursor.getInt(INDEX_ERROR_CODE));
        item.setLocked(cursor.getInt(INDEX_LOCKED) != 0);
        item.setPerson(cursor.getInt(INDEX_PERSON));
        item.setProtocol(cursor.getInt(INDEX_PROTOCOL));
        item.setRead(cursor.getInt(INDEX_READ) != 0);
        item.setSeen(cursor.getInt(INDEX_SEEN) != 0);
        item.setStatus(cursor.getInt(INDEX_STATUS));
        item.setSubject(empty(cursor, INDEX_SUBJECT));
        item.setThreadId(cursor.getLong(INDEX_THREAD_ID));
        item.setType(cursor.getInt(INDEX_TYPE));
    }

    @Override
    public String[] getReadPermissions() {
        return PERMISSIONS_READ;
    }

    @Override
    public String[] getDeletePermissions() {
        return PERMISSIONS_WRITE;
    }
}

File: duplicates-android/app/src/main/java/com/github/duplicates/message/MessageProvider.kt
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.message

import android.Manifest
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.provider.BaseColumns._ID
import android.provider.Telephony
import android.provider.Telephony.TextBasedSmsColumns.*
import com.github.duplicates.DuplicateProvider

/**
 * Provide duplicate messages.
 *
 * @author moshe.w
 */
class MessageProvider(context: Context) : DuplicateProvider<MessageItem>(context) {

    override fun getContentUri(): Uri? {
        return Telephony.Sms.CONTENT_URI
    }

    override fun getCursorProjection(): Array<String>? {
        return PROJECTION
    }

    override fun createItem(cursor: Cursor): MessageItem? {
        return MessageItem()
    }

    override fun populateItem(cursor: Cursor, item: MessageItem) {
        item.id = cursor.getLong(INDEX_ID)
        item.address = empty(cursor, INDEX_ADDRESS)
        item.body = empty(cursor, INDEX_BODY)
        item.dateReceived = cursor.getLong(INDEX_DATE_RECEIVED)
        item.dateSent = cursor.getLong(INDEX_DATE_SENT)
        item.errorCode = cursor.getInt(INDEX_ERROR_CODE)
        item.isLocked = cursor.getInt(INDEX_LOCKED) != 0
        item.person = cursor.getInt(INDEX_PERSON)
        item.protocol = cursor.getInt(INDEX_PROTOCOL)
        item.isRead = cursor.getInt(INDEX_READ) != 0
        item.isSeen = cursor.getInt(INDEX_SEEN) != 0
        item.status = cursor.getInt(INDEX_STATUS)
        item.subject = empty(cursor, INDEX_SUBJECT)
        item.threadId = cursor.getLong(INDEX_THREAD_ID)
        item.type = cursor.getInt(INDEX_TYPE)
    }

    override fun getReadPermissions(): Array<String>? {
        return PERMISSIONS_READ
    }

    override fun getDeletePermissions(): Array<String>? {
        return PERMISSIONS_WRITE
    }

    companion object {

        private val PERMISSIONS_READ = arrayOf(Manifest.permission.READ_SMS)
        private val PERMISSIONS_WRITE = arrayOf("android.permission.WRITE_SMS")

        private val PROJECTION = arrayOf(
            _ID,
            ADDRESS,
            BODY,
            DATE,
            DATE_SENT,
            ERROR_CODE,
            LOCKED,
            PERSON,
            PROTOCOL,
            READ,
            SEEN,
            STATUS,
            SUBJECT,
            THREAD_ID,
            TYPE
        )

        private const val INDEX_ID = 0
        private const val INDEX_ADDRESS = 1
        private const val INDEX_BODY = 2
        private const val INDEX_DATE_RECEIVED = 3
        private const val INDEX_DATE_SENT = 4
        private const val INDEX_ERROR_CODE = 5
        private const val INDEX_LOCKED = 6
        private const val INDEX_PERSON = 7
        private const val INDEX_PROTOCOL = 8
        private const val INDEX_READ = 9
        private const val INDEX_SEEN = 10
        private const val INDEX_STATUS = 11
        private const val INDEX_SUBJECT = 12
        private const val INDEX_THREAD_ID = 13
        private const val INDEX_TYPE = 14
    }
}
